llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -677,6 +693,33 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
     return !A.checkForAllCallLikeInstructions(DoesNotRetrieve, *this,
                                               UsedAssumedInformation);
   }
+
+  // Returns true if FlatScratchInit is needed, i.e., no-flat-scratch-init is
+  // not to be set.
+  bool needFlatScratchInit(Attributor &A) {
+    // This is called on each callee; false means callee shouldn't have
+    // no-flat-scratch-init.
+    auto CheckForNoFlatScratchInit = [&](Instruction &I) {
+      const auto &CB = cast<CallBase>(I);
+      const Value *CalleeOp = CB.getCalledOperand();
+      const Function *Callee = dyn_cast<Function>(CalleeOp);
+      if (!Callee) // indirect call
+        return CB.isInlineAsm();
```

Comment:
You should replace all this handling, starting from here till the end of this lambda, with a call to 
```
 /// Check \p Pred on all potential Callees of \p CB.
  ///
  /// This method will evaluate \p Pred with all potential callees of \p CB as
  /// input and return true if \p Pred does. If some callees might be unknown
  /// this function will return false.
  bool checkForAllCallees(
      function_ref<bool(ArrayRef<const Function *> Callees)> Pred,
      const AbstractAttribute &QueryingAA, const CallBase &CB);
```
This will work for indirect calls as well.

Reply:
I'm not sure I follow. This `checkForAllCallees()` function requires a pred. So we wrap the existing code of the lambda in a new lambda, and pass the new lambda to `checkFroAllCallees()`? Also, could you pls explain a little what the potential problem is without using this function?

Reply:
@jdoerfert ping here.

Reply:
Instead of checking for the call instructions with checkForAllCallLikeInstructions, you instead use checkForAllCallees for all direct and possible indirect call candidates (not 100% sure how this interacts with asm). That way you don't have to concern yourself with indirect calls, since any possible call targets should also appear in the callees

Reply:
`checkForAllCallees()` requires an arg which is a `CallBase`, i.e., a call-like instruction. So, to use it, you need to get the call instruction first. On the other hand, `checkForAllCallLikeInstructions()` can give you all the call instructions. It's not clear to me how `checkForAllCallees` can replace `checkForAllCallLikeInstructions`.

Reply:
`checkForAllCallees()` seems fairly new. If I understand it correctly, we will need to move most of this lambda to a call-site attribute as a helper to do all the real work. Then `needFlatScratchInit()` will retrieve that attribute on every call-like instruction. Is that correct? It definitely looks cleaner that way, I think. But then, the same could be done for all the other function attributes being tracked here? Should that be captured as a pattern in the attributor itself? "a function attribute that is computed by examining the same attribute for every callsite".

Reply:
@ssahasra I fully agree with your description of how it works if `checkForAllCallees()` is used. Based on this, I'm not sure I see the advantage of using `checkForAllCallees()`. I think you still need to check all call-like instructions, and all it does is to move the real work (check for intrinsics, indirect calls etc) to this CallSite attribute. Furthermore, don't you need both a CallSite attribute and a Function attribute, because when you set up the CallSite attribute you need to retrieve some attribute from the callee? Besides, I want to mention that I was not able to find any use of `checkForAllCallees()` or any definition of CallSite attributes in the existing code.
@arsenm @jdoerfert Your thoughts?

Reply:
This is backwards. checkForAllCallees's callback takes only the candidate callee function, not the callsite. The point is it passes possible indirect callees you cannot trivially infer from just the callsite. This does mean you can miss manually specified callsite attributes, but that rarely matters.

Just use checkForAllCallees.

Reply:
> This is backwards. checkForAllCallees's callback takes only the candidate callee function, not the callsite.

Ummm, no it does not? Unless we are looking at different things, it takes a `CallBase`:

```
   1737 bool Attributor::checkForAllCallees(
   1738     function_ref<bool(ArrayRef<const Function *>)> Pred,
   1739     const AbstractAttribute &QueryingAA, const CallBase &CB) {
```

> The point is it passes possible indirect callees you cannot trivially infer from just the callsite. This does mean you can miss manually specified callsite attributes, but that rarely matters.

Yes, that's the advantage.
 
> Just use checkForAllCallees.

But how exactly? When computing a function attribute, we have to do `getAAFor()` on every callsite in that function, right?

Reply:
> I was not able to find any use of `checkForAllCallees()` or any definition of CallSite attributes in the existing code. 

It is used through the `AACalleeToCallSite` helper class. This class is used by a lot of attributes that are computed over the callgraph. Take a look at how `AANoFree` is constructed for example. It is declared as both a function attribute and callsite attribute.

@arsenm @jdoerfert Your thoughts?

Reply:
If we were to use `callForAllCallees()`, based on my understanding, the code would look something like this:
```diff 
   bool needFlatScratchInit(Attributor &A) {
     // This is called on each callee; false means callee shouldn't have
     // no-flat-scratch-init.
     auto CheckForNoFlatScratchInit = [&](Instruction &I) {
       const auto &CB = cast<CallBase>(I);
-      const Value *CalleeOp = CB.getCalledOperand();
-      const Function *Callee = dyn_cast<Function>(CalleeOp);
-      if (!Callee) // indirect call
-        return CB.isInlineAsm();
-
-      if (Callee->isIntrinsic())
-        return Callee->getIntrinsicID() != Intrinsic::amdgcn_addrspacecast_nonnull;
-
-      const auto *CalleeInfo = A.getAAFor<AAAMDAttributes>(
-          *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
-      return CalleeInfo && CalleeInfo->isAssumed(FLAT_SCRATCH_INIT);
+      auto PredOnCallees = [&](ArrayRef<const Function *> Callees) {
+        bool Ret = true;
+        for (const Function *Callee : Callees) {
+          if (!Callee) { // indirect call
+            // non-asm indirect call is already handled in updateImpl()
+            assert(CB.isInlineAsm());
+            continue;
+          }
+
+          if (Callee->isIntrinsic()) {
+            Ret &= (Callee->getIntrinsicID() != Intrinsic::amdgcn_addrspacecast_nonnull);
+            if (!Ret)
+              return false;
+            continue;
+          }
+
+          const auto *CalleeInfo = A.getAAFor<AAAMDAttributes>(
+              *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
+          Ret &= (CalleeInfo && CalleeInfo->isAssumed(FLAT_SCRATCH_INIT));
+          if (!Ret)
+            return false;
+        }
+        return Ret;
+      };
+
+      return A.checkForAllCallees(PredOnCallees, *this, CB);
     };

     bool UsedAssumedInformation = false;
    // If any callee is false (i.e. need FlatScratchInit),
    // checkForAllCallLikeInstructions returns false, in which case this
    // function returns true.
    return !A.checkForAllCallLikeInstructions(CheckForNoFlatScratchInit, *this,
                                              UsedAssumedInformation);
  }
```
Here, in the lambda `CheckForNoFlatScratchInit`, a new lambda is created (`PredOnCallees`), and it simply calls `checkForAllCallees()` passing the new lambda. 
@arsenm @jdoerfert Pls let me know if this is what you had in mind.

Reply:
I assume you shouldn't need to handle a null callee in the callback. Also, I don't think you have the CallBase here.

It's easier to review an actual patch, so submit a new one using checkForAllCallees?

Reply:
Code has been updated after fixing a bug regarding indirect calls of known callees. I'm not sure using `checkForAllCallees()` is necessary for 2 reasons: (1) code is now very simple, boiling down to just checking for intrinsics (2) the first part of `AAAMDAttributesFunction::updateImpl()` already does similar work as `checkForAllCallees()`:
```cpp
  ChangeStatus updateImpl(Attributor &A) override {
    ...
    // Check for Intrinsics and propagate attributes.
    const AACallEdges *AAEdges = A.getAAFor<AACallEdges>(
        *this, this->getIRPosition(), DepClassTy::REQUIRED);
    if (!AAEdges || AAEdges->hasNonAsmUnknownCallee())
      return indicatePessimisticFixpoint();
...
    for (Function *Callee : AAEdges->getOptimisticEdges()) { // Jun - what checkForAllCallees() does is essentially calling getOptimisticEdges() and then running the pred on the edges
      Intrinsic::ID IID = Callee->getIntrinsicID();
      if (IID == Intrinsic::not_intrinsic) {
        const AAAMDAttributes *AAAMD = A.getAAFor<AAAMDAttributes>(
            *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
        if (!AAAMD)
          return indicatePessimisticFixpoint();
        *this &= *AAAMD;
        continue;
      }
...
```

Reply:
On the other hand, a patch where `checkForAllCallees()` is used is as follows:
```diff
   bool needFlatScratchInit(Attributor &A) {
     assert(isAssumed(FLAT_SCRATCH_INIT)); // only called if the bit is still set
     auto CheckForNoFlatScratchInit = [&](Instruction &I) {
       const auto &CB = cast<CallBase>(I);
-      const Function *Callee = CB.getCalledFunction();
-
-      if (Callee && Callee->isIntrinsic())
-        return Callee->getIntrinsicID() !=
-               Intrinsic::amdgcn_addrspacecast_nonnull;
-
+      auto PredOnCallees = [&](ArrayRef<const Function *> Callees) {
+        for (const Function *Callee : Callees) {
+          if (Callee->isIntrinsic())
+            if (Callee->getIntrinsicID() == Intrinsic::amdgcn_addrspacecast_nonnull)
+              return false;
+        }
+        return true;
+      };
+      if(!A.checkForAllCallees(PredOnCallees, *this, CB))
+        return CB.isInlineAsm();
       return true;
     };

     bool UsedAssumedInformation = false;
     return !A.checkForAllCallLikeInstructions(CheckForNoFlatScratchInit, *this,
```

Reply:
The point of using chekcForAllCallees is it also works for indirect calls, not that it makes the code simpler or better. You can do that in a follow up patch if you really want, but you should still use it